lib/notifications/display.dart

```diff
@@ -395,4 +371,60 @@ class NotificationDisplayManager {
     }
     return null;
   }
+
+  /// Generates a URI from the FCM message data, embedding the [narrowLink]
+  /// and the userID.
+  ///
+  /// This URI is used to route to the relevant conversation's [MessageListPage]
+  /// when a notification with this intent URI is opened.
+  @visibleForTesting
+  static Future<Uri?> notificationIntentUriFromFcmMessage({
+    required GlobalStore globalStore,
+    required MessageFcmMessage data,
+  }) async {
+    final account = globalStore.accounts.firstWhereOrNull((account) =>
+      account.realmUrl.origin == data.realmUri.origin && account.userId == data.userId);
+    if (account == null) {
+      assert(debugLog('notif: failed to find a matching account: realmUri=${data.realmUri} userId=${data.userId}')); // TODO(log)
+      return null;
+    }
+
+    final narrow = switch (data.recipient) {
+      FcmMessageChannelRecipient(:var streamId, :var topic) =>
+        TopicNarrow(streamId, topic),
+      FcmMessageDmRecipient(:var allRecipientIds) =>
+        DmNarrow(allRecipientIds: allRecipientIds, selfUserId: account.userId),
+    };
+    final perAccountStore = await globalStore.perAccount(account.id);
+    final narrowUri = narrowLink(perAccountStore, narrow, nearMessageId: data.zulipMessageId);
```

Comment:
nit: use just `store` for the name of a PerAccountStore local variable
```suggestion
    final store = await globalStore.perAccount(account.id);
    final narrowUri = narrowLink(store, narrow, nearMessageId: data.zulipMessageId);
```

That's the convention we consistently use across the codebase. Then when there is a GlobalStore, it's distinguished explicitly as `globalStore`, even if there doesn't happen to be a PerAccountStore in scope at the same spot.